File: windows/tests/ConnectorControl.Core.Tests/State/AppStateCommandTests.cs
using ConnectorControl.Core.State;
using ConnectorControl.Core.Tests.TestSupport;

namespace ConnectorControl.Core.Tests.State;

public class AppStateCommandTests
{
    [Fact]
    public void QuitAsksForConfirmationByDefault()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var quit = 0;
        state.QuitRequested += () => quit++;
        h.Dialogs.NextConfirm = false;
        state.QuitApp();
        Assert.Equal(0, quit);
        Assert.Equal(new FakeDialogs.ConfirmCall("Quit Connector Control?", null, "Quit", "Cancel", false), h.Dialogs.Confirms[0]);
        h.Dialogs.NextConfirm = true;
        state.QuitApp();
        Assert.Equal(1, quit);
    }

    [Fact]
    public void QuitSkipsTheConfirmationWhenDisabled()
    {
        using var h = new AppStateHarness();
        h.Settings.ConfirmBeforeQuit = false;
        using var state = h.Create();
        var quit = 0;
        state.QuitRequested += () => quit++;
        state.QuitApp();
        Assert.Equal(1, quit);
        Assert.Empty(h.Dialogs.Confirms);
    }

    [Fact]
    public async Task RestartClaudeConfirmsThenRestarts()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        await state.RestartClaudeAsync();
        Assert.Equal(new FakeDialogs.ConfirmCall("Restart Claude Desktop now?", "Any in-progress Claude conversation will be interrupted.", "Restart", "Cancel", false), h.Dialogs.Confirms[0]);
        Assert.Equal(1, h.Claude.RestartCalls);
    }

    [Fact]
    public async Task RestartClaudeCancelledDoesNothing()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        h.Dialogs.NextConfirm = false;
        await state.RestartClaudeAsync();
        Assert.Equal(0, h.Claude.RestartCalls);
    }

    [Fact]
    public async Task RestartClaudeSkipsTheConfirmationWhenDisabled()
    {
        using var h = new AppStateHarness();
        h.Settings.ConfirmBeforeRestart = false;
        using var state = h.Create();
        await state.RestartClaudeAsync();
        Assert.Empty(h.Dialogs.Confirms);
        Assert.Equal(1, h.Claude.RestartCalls);
    }

    [Fact]
    public async Task RestartErrorLandsInLastErrorAndARecheckIsScheduled()
    {
        using var h = new AppStateHarness();
        h.Settings.ConfirmBeforeRestart = false;
        h.Settings.LastApplyDate = h.Now;
        h.Claude.IsRunning = true;
        h.Claude.LaunchDate = h.Now.AddHours(-1);
        using var state = h.Create();
        Assert.True(state.NeedsClaudeRestart);
        h.Claude.RestartResult = "Claude didn’t quit (it may be showing a dialog). Quit it manually, then click Restart Claude again.";
        await state.RestartClaudeAsync();
        Assert.Equal(1, h.Ui.Pending);   // the completion is posted through the host, not run inline
        h.Ui.Pump();
        Assert.Equal(h.Claude.RestartResult, state.LastError);
        Assert.True(state.NeedsClaudeRestart);
        Assert.Equal([TimeSpan.FromSeconds(3), TimeSpan.FromSeconds(20)], h.Delays.Pending.Select(d => d.Delay).ToArray());
        h.Claude.IsRunning = false;   // the user quit it by hand in the meantime
        h.Delays.RunNext();
        Assert.False(state.NeedsClaudeRestart);
        h.Delays.RunNext();           // the 20 s look: LastError already says what went wrong, so leave it
        Assert.Equal(h.Claude.RestartResult, state.LastError);
    }

    /// <summary>C#-only: the Mac's restart reports through a completion handler and cannot throw,
    /// while Windows awaits a Task that can.</summary>
    [Fact]
    public async Task RestartExceptionOutsideTheLaunchGuardStillCompletesWithAMessage()
    {
        using var h = new AppStateHarness();
        h.Settings.ConfirmBeforeRestart = false;
        using var state = h.Create();
        h.Claude.OnRestart = () => throw new InvalidOperationException("boom");
        await state.RestartClaudeAsync();
        h.Ui.Pump();
        Assert.Equal(1, h.Claude.RestartCalls);
        Assert.Equal("boom", state.LastError);
        // Same schedule as the success path: the marshalled completion always runs to the end.
        Assert.Equal([TimeSpan.FromSeconds(3), TimeSpan.FromSeconds(20)], h.Delays.Pending.Select(d => d.Delay).ToArray());
    }

    /// <summary>C#-only: Windows relaunches through explorer.exe, which reports success before Claude
    /// appears, so it looks again 20 s later; the Mac's restarter launches Claude itself and
    /// reports a failed launch through its completion.</summary>
    [Fact]
    public async Task ARelaunchThatSilentlyFailedIsReportedTwentySecondsLater()
    {
        using var h = new AppStateHarness();
        h.Settings.ConfirmBeforeRestart = false;
        h.Settings.LastApplyDate = h.Now;
        h.Claude.IsRunning = true;
        h.Claude.LaunchDate = h.Now.AddHours(-1);
        using var state = h.Create();
        // RestartAsync reports success: explorer.exe accepted the AUMID. Claude never appeared.
        h.Claude.OnRestart = () =>
        {
            h.Claude.IsRunning = false;
            h.Claude.LaunchDate = null;
        };
        await state.RestartClaudeAsync();
        h.Ui.Pump();
        Assert.Null(state.LastError);
        Assert.Equal(AppState.RestartRecheckDelay, h.Delays.Pending[0].Delay);
        Assert.Equal(AppState.RestartRelaunchCheck, h.Delays.Pending[1].Delay);

        h.Delays.RunNext();           // 3 s: too early to conclude anything
        Assert.Null(state.LastError);
        h.Delays.RunNext();           // 20 s: the probe's relaunch bound has passed
        Assert.Equal(AppState.RelaunchFailedMessage, state.LastError);
        Assert.Equal("Claude didn’t come back after the restart. Start Claude yourself, then try again.", AppState.RelaunchFailedMessage);
    }

    [Fact]
    public async Task RestartSuccessClearsTheErrorAndTheRestartState()
    {
        using var h = new AppStateHarness();
        h.Settings.ConfirmBeforeRestart = false;
        h.Settings.LastApplyDate = h.Now;
        h.Claude.IsRunning = true;
        h.Claude.LaunchDate = h.Now.AddHours(-1);
        using var state = h.Create();
        state.LastError = "old banner";
        h.Claude.OnRestart = () => h.Claude.LaunchDate = h.Now.AddSeconds(1);
        await state.RestartClaudeAsync();
        h.Ui.Pump();
        Assert.Null(state.LastError);
        Assert.False(state.NeedsClaudeRestart);
    }

    [Fact]
    public void ToastRestartActionIsGuardedByAPendingRestartAndSkipsTheConfirmation()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        h.Notifier.ActivateRestart();   // stale click: nothing pending
        Assert.Equal(0, h.Claude.RestartCalls);

        h.Claude.IsRunning = true;
        h.Claude.LaunchDate = h.Now.AddHours(-1);
        state.SetEnabled("aws-mcp", false);
        Assert.True(state.NeedsClaudeRestart);
        h.Notifier.ActivateRestart();
        Assert.Equal(1, h.Claude.RestartCalls);
        Assert.Empty(h.Dialogs.Confirms);   // the explicit action click IS the confirmation
    }

    [Fact]
    public void DisposeUnsubscribesTheToastRestartAction()
    {
        using var h = new AppStateHarness();
        var state = h.Create();
        h.Claude.IsRunning = true;
        h.Claude.LaunchDate = h.Now.AddHours(-1);
        state.SetEnabled("aws-mcp", false);
        Assert.True(state.NeedsClaudeRestart);   // a pending restart, so ActivateRestart would act if still wired up
        state.Dispose();
        h.Notifier.ActivateRestart();
        Assert.Equal(0, h.Claude.RestartCalls);
    }

    [Fact]
    public async Task PendingRestartDelaysAreNoOpsAfterDispose()
    {
        using var h = new AppStateHarness();
        h.Settings.ConfirmBeforeRestart = false;
        h.Settings.LastApplyDate = h.Now;
        h.Claude.IsRunning = true;
        h.Claude.LaunchDate = h.Now.AddHours(-1);
        var state = h.Create();
        await state.RestartClaudeAsync();
        h.Ui.Pump();
        Assert.Equal(2, h.Delays.Pending.Count);
        var errorBefore = state.LastError;
        var needsRestartBefore = state.NeedsClaudeRestart;
        h.Claude.IsRunning = false;   // without the guard the 20 s check would now report a failed relaunch
        state.Dispose();
        h.Delays.RunNext();   // 3 s recheck: must be a no-op post-Dispose, not throw
        h.Delays.RunNext();   // 20 s relaunch check: same
        Assert.Equal(errorBefore, state.LastError);
        Assert.Equal(needsRestartBefore, state.NeedsClaudeRestart);

        // The completion block itself — not only the two delayed follow-ups — must be a no-op
        // once disposed: a restart that finishes after Dispose (the app quitting mid-restart)
        // must not resurrect state or schedule fresh delays.
        state.LastError = "should not survive";
        await state.PerformRestartClaudeAsync();
        h.Ui.Pump();
        Assert.Equal("should not survive", state.LastError);
        Assert.Empty(h.Delays.Pending);   // no new delays scheduled on a disposed state
    }

    [Fact]
    public void SwitchCollectionAppliesImmediately()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        Assert.Null(state.CreateCollection("Work"));
        Assert.Equal("Work", state.ActiveCollection);
        state.SetEnabled("aws-mcp", false);
        Assert.Equal(["scoutbook", "service-now"], AppStateHarness.Keys(h.ClaudeServers().Keys));

        h.Settings.LastApplyDate = null;
        state.SwitchCollection("Default");
        Assert.Equal("Default", state.ActiveCollection);
        Assert.Equal(["aws-mcp", "scoutbook", "service-now"], AppStateHarness.Keys(h.ClaudeServers().Keys));
        Assert.Equal(h.Now, h.Settings.LastApplyDate);
        Assert.Equal("Default", h.StoreOnDisk().ActiveCollection);
    }

    [Fact]
    public void SwitchCollectionIgnoresAnUnknownName()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        state.SwitchCollection("Nope");
        Assert.Equal("Default", state.ActiveCollection);
        Assert.Null(state.LastError);
        Assert.Null(h.Settings.LastApplyDate);
    }
}

File: Tests/ConnectorControlStateTests/AppStateCommandTests.swift
import XCTest
import ConnectorControlCore
@testable import ConnectorControlState

/// windows/tests/ConnectorControl.Core.Tests/State/AppStateCommandTests.cs. The
/// restart completion is posted to the marshal queue, so every restart test
/// pumps once where the C# awaited and pumped.
@MainActor
final class AppStateCommandTests: XCTestCase {
    private let fixture = ["aws-mcp", "scoutbook", "service-now"]

    func testQuitAsksForConfirmationByDefault() {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        var quit = 0
        state.quitRequested = { quit += 1 }
        h.dialogs.nextConfirm = false
        state.quitApp()
        XCTAssertEqual(quit, 0)
        XCTAssertEqual(h.dialogs.confirms[0], FakeDialogs.ConfirmCall(
            message: "Quit Connector Control?", informative: nil, primary: "Quit", cancel: "Cancel", destructive: false))
        h.dialogs.nextConfirm = true
        state.quitApp()
        XCTAssertEqual(quit, 1)
    }

    func testQuitSkipsTheConfirmationWhenDisabled() {
        let h = AppStateHarness()
        defer { h.dispose() }
        h.settings.confirmBeforeQuit = false
        let state = h.create()
        var quit = 0
        state.quitRequested = { quit += 1 }
        state.quitApp()
        XCTAssertEqual(quit, 1)
        XCTAssertTrue(h.dialogs.confirms.isEmpty)
    }

    func testRestartClaudeConfirmsThenRestarts() {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        state.restartClaude()
        XCTAssertEqual(h.dialogs.confirms[0], FakeDialogs.ConfirmCall(
            message: "Restart Claude Desktop now?",
            informative: "Any in-progress Claude conversation will be interrupted.",
            primary: "Restart", cancel: "Cancel", destructive: false))
        XCTAssertEqual(h.claude.restartCalls, 1)
    }

    func testRestartClaudeCancelledDoesNothing() {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        h.dialogs.nextConfirm = false
        state.restartClaude()
        XCTAssertEqual(h.claude.restartCalls, 0)
    }

    func testRestartClaudeSkipsTheConfirmationWhenDisabled() {
        let h = AppStateHarness()
        defer { h.dispose() }
        h.settings.confirmBeforeRestart = false
        let state = h.create()
        state.restartClaude()
        XCTAssertTrue(h.dialogs.confirms.isEmpty)
        XCTAssertEqual(h.claude.restartCalls, 1)
    }

    func testRestartErrorLandsInLastErrorAndARecheckIsScheduled() {
        let h = AppStateHarness()
        defer { h.dispose() }
        h.settings.confirmBeforeRestart = false
        h.settings.lastApplyDate = h.now
        h.claude.isRunning = true
        h.claude.launchDate = h.now.addingTimeInterval(-3600)
        let state = h.create()
        XCTAssertTrue(state.needsClaudeRestart)
        h.claude.restartResult = "Claude didn’t quit (it may be showing a dialog). Quit it manually, then click Restart Claude again."
        state.restartClaude()
        XCTAssertEqual(h.ui.pending, 1, "the completion is posted through the host, not run inline")
        h.ui.pump()
        XCTAssertEqual(state.lastError, h.claude.restartResult)
        XCTAssertTrue(state.needsClaudeRestart)
        XCTAssertEqual(h.delays.pending.map(\.delay), [3])
        h.claude.isRunning = false   // the user quit it by hand in the meantime
        h.delays.runNext()
        XCTAssertFalse(state.needsClaudeRestart)
    }

    func testRestartSuccessClearsTheErrorAndTheRestartState() {
        let h = AppStateHarness()
        defer { h.dispose() }
        h.settings.confirmBeforeRestart = false
        h.settings.lastApplyDate = h.now
        h.claude.isRunning = true
        h.claude.launchDate = h.now.addingTimeInterval(-3600)
        let state = h.create()
        state.lastError = "old banner"
        h.claude.onRestart = { h.claude.launchDate = h.now.addingTimeInterval(1) }
        state.restartClaude()
        h.ui.pump()
        XCTAssertNil(state.lastError)
        XCTAssertFalse(state.needsClaudeRestart)
    }

    func testNotificationRestartActionIsGuardedByAPendingRestartAndSkipsTheConfirmation() {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        h.notifier.activateRestart()   // stale click: nothing pending
        XCTAssertEqual(h.claude.restartCalls, 0)

        h.claude.isRunning = true
        h.claude.launchDate = h.now.addingTimeInterval(-3600)
        state.setEnabled("aws-mcp", false)
        XCTAssertTrue(state.needsClaudeRestart)
        h.notifier.activateRestart()
        XCTAssertEqual(h.claude.restartCalls, 1)
        XCTAssertTrue(h.dialogs.confirms.isEmpty)   // the explicit action click IS the confirmation
    }

    func testDisposeUnsubscribesTheNotificationRestartAction() {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        h.claude.isRunning = true
        h.claude.launchDate = h.now.addingTimeInterval(-3600)
        state.setEnabled("aws-mcp", false)
        XCTAssertTrue(state.needsClaudeRestart)   // a pending restart, so activateRestart would act if still wired up
        state.dispose()
        h.notifier.activateRestart()
        XCTAssertEqual(h.claude.restartCalls, 0)
    }

    func testPendingRestartDelaysAreNoOpsAfterDispose() {
        let h = AppStateHarness()
        defer { h.dispose() }
        h.settings.confirmBeforeRestart = false
        h.settings.lastApplyDate = h.now
        h.claude.isRunning = true
        h.claude.launchDate = h.now.addingTimeInterval(-3600)
        let state = h.create()
        state.restartClaude()
        h.ui.pump()
        XCTAssertEqual(h.delays.pending.count, 1)
        let errorBefore = state.lastError
        let needsRestartBefore = state.needsClaudeRestart
        h.claude.isRunning = false   // without the guard the recheck would now clear needsClaudeRestart
        state.dispose()
        h.delays.runNext()   // 3 s recheck: must be a no-op post-dispose, not a crash
        XCTAssertEqual(state.lastError, errorBefore)
        XCTAssertEqual(state.needsClaudeRestart, needsRestartBefore)
    }

    func testSwitchProfileAppliesImmediately() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        h.dialogs.nextPromptAnswer = "Work"
        state.newProfile()
        XCTAssertEqual(state.activeProfile, "Work")
        state.setEnabled("aws-mcp", false)
        XCTAssertEqual(try h.claudeServers().keys.sorted(), ["scoutbook", "service-now"])

        h.settings.lastApplyDate = nil
        state.switchProfile(to: "Default")
        XCTAssertEqual(state.activeProfile, "Default")
        XCTAssertEqual(try h.claudeServers().keys.sorted(), fixture)
        XCTAssertEqual(h.settings.lastApplyDate, h.now)
        XCTAssertEqual(try h.storeOnDisk().activeProfile, "Default")
    }

    func testSwitchProfileIgnoresAnUnknownName() {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        state.switchProfile(to: "Nope")
        XCTAssertEqual(state.activeProfile, "Default")
        XCTAssertNil(state.lastError)
        XCTAssertNil(h.settings.lastApplyDate)
    }

    func testNewProfilePromptTextAndCancel() {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        h.dialogs.nextPromptAnswer = nil
        state.newProfile()
        XCTAssertEqual(h.dialogs.prompts[0], FakeDialogs.PromptCall(title: "New Profile", initial: ""))
        XCTAssertEqual(state.profileNames, ["Default"])

        h.dialogs.nextPromptAnswer = "Work"
        state.newProfile()
        XCTAssertEqual(state.profileNames, ["Default", "Work"])
        XCTAssertEqual(state.sortedNames, fixture)   // a COPY of the active profile
        XCTAssertEqual(h.settings.lastApplyDate, h.now)
    }

    func testNewProfileErrorsGoToLastError() {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        h.dialogs.nextPromptAnswer = "Default"
        state.newProfile()
        XCTAssertEqual(state.lastError, "A profile named “Default” already exists.")
        h.dialogs.nextPromptAnswer = "   "
        state.newProfile()
        XCTAssertEqual(state.lastError, "Name must not be empty.")
        XCTAssertEqual(state.profileNames, ["Default"])
    }

    func testRenameProfilePrefillsTheActiveName() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        h.dialogs.nextPromptAnswer = "Main"
        state.renameProfile()
        XCTAssertEqual(h.dialogs.prompts[0], FakeDialogs.PromptCall(title: "Rename Profile", initial: "Default"))
        XCTAssertEqual(state.activeProfile, "Main")
        XCTAssertEqual(try h.storeOnDisk().activeProfile, "Main")
        XCTAssertEqual(h.settings.lastApplyDate, h.now)
    }

    func testDeleteProfileConfirmTextAndLastProfileError() {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        state.deleteProfile()
        XCTAssertEqual(h.dialogs.confirms[0], FakeDialogs.ConfirmCall(
            message: "Delete Profile “Default”?",
            informative: "Its connector list is removed; backups keep prior states.",
            primary: "Delete", cancel: "Cancel", destructive: true))
        XCTAssertEqual(state.lastError, "Can’t delete the last profile.")
        XCTAssertEqual(state.profileNames, ["Default"])
    }

    func testDeleteProfileSwitchesToTheAlphabeticallyFirstRemaining() throws {
        let h = AppStateHarness()
        defer { h.dispose() }
        let state = h.create()
        h.dialogs.nextPromptAnswer = "Zeta"
        state.newProfile()
        h.dialogs.nextPromptAnswer = "Work"
        state.newProfile()
        XCTAssertEqual(state.activeProfile, "Work")
        h.dialogs.nextConfirm = false
        state.deleteProfile()
        XCTAssertEqual(state.profileNames, ["Default", "Work", "Zeta"])   // cancelled
        h.dialogs.nextConfirm = true
        state.deleteProfile()
        XCTAssertEqual(state.profileNames, ["Default", "Zeta"])
        XCTAssertEqual(state.activeProfile, "Default")
        XCTAssertEqual(try h.storeOnDisk().activeProfile, "Default")
    }
}
